### DISCUSSION TITLE
STM32 Timer: connect and configureOutputChannel

### DISCUSSION THREAD

**Discussion by strongly-typed** [modm-io/modm#908] [Ideas]
`connect` uses a symbolic template parameter to connect to the pin:

[`Timer4::connect<GpioB7::Ch2>();`](https://github.com/modm-io/modm/blob/develop/examples/blue_pill_f103/servo_pwm/main.cpp#L40)

whereas `configureOutputChannel` uses a plain int for this.

[`Timer4::configureOutputChannel(2, Timer4::OutputCompareMode::Pwm, 0);`](https://github.com/modm-io/modm/blob/develop/examples/blue_pill_f103/servo_pwm/main.cpp#L40)

I would love to create a template class that gets `GpioB7::Ch2` as a template parameter and then uses this for `connect` (possible) *and* deduced the channel number `2` from the template parameter `GpioB7::Ch2`.

Is there any chance to get something like `GpioB7::Ch2::ChId` or similar that can be passed to `configureOutputChannel`?

**Comment by chris-durand** [comment]
I have made a proof-of-concept on [that branch](https://github.com/chris-durand/modm/tree/feature/stm32_timer_channel_signal).

With that change you can now also write:
```
Timer4::configureOutputChannel<GpioB7::Ch2>(Timer4::OutputCompareMode::Pwm, 0);
Timer4::setCompareValue<GpioB7::Ch2>(1000);
```
This only implements it for those two methods in the general-purpose timer implementation but feel free to take my changes and extend them if you like.

**Comment by salkinium** [comment]
Nice, though it would have to be for backward compatibility:
```cpp
Timer4::configureOutputSignal<GpioB7::Ch2>(Timer4::OutputCompareMode::Pwm, 0);
Timer4::setCompareValueSignal<GpioB7::Ch2>(1000);
```
or alternatively you could also do this less elegant version:
```cpp
Timer4::configureOutputChannel(Timer4::signalToChannel<GpioB7::Ch2>(), Timer4::OutputCompareMode::Pwm, 0);
```

**Reply by chris-durand** [reply]
> Nice, though it would have to be for backward compatibility:
> 
> ```c++
> Timer4::configureOutputSignal<GpioB7::Ch2>(Timer4::OutputCompareMode::Pwm, 0);
> Timer4::setCompareValueSignal<GpioB7::Ch2>(1000);
> ```
> 
> or alternatively you could also do this less elegant version:
> 
> ```c++
> Timer4::configureOutputChannel(Timer4::signalToChannel<GpioB7::Ch2>(), Timer4::OutputCompareMode::Pwm, 0);
> ```

I don't see why. The solution I proposed should have zero issues with backward compatibility. You have two overloads of which one is a template taking the signal and the other one isn't.

**Reply by salkinium** [reply]
Oh, you're right! The number of arguments is different too.

**Reply by strongly-typed** [reply]
I like the solution. Can we get this merged soon?

**Reply by salkinium** [reply]
Open a PR and I'll merge it…

**Reply by chris-durand** [reply]
This solution is mergeable, but not complete. It would be nice if it covered more API functions and also the Advanced Control Timers. All of that could be added later if you need it merged now.

**Comment by salkinium** [comment]
This feature has been implemented in https://github.com/modm-io/modm/pull/912
